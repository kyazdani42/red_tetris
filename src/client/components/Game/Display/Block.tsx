import * as React from 'react';

export const NextBlock = (color: string, key: string) => (
  <div
    style={{
      backgroundColor: changeBlockColor(color),
      border: '1px solid black',
      width: 'calc(25% - 2px)',
      height: 'calc(25% - 2px)'
    }}
    key={key}
  />
);

export const Block = (color: string, key: string) => (
  <div
    style={{
      backgroundColor: changeBlockColor(color),
      border: '1px solid black',
      width: 'calc(10% - 2px)',
      height: 'calc(5% - 2px)'
    }}
    key={key}
  />
);

export const changeBlockColor = (color: string) => {
  switch (color) {
    case 'skyBlue':
     return '#adf0ff';
    case 'blue':
      return '#7427CC';
    case 'orange':
      return '#EF6A00';
    case 'yellow':
      return '#FFC851';
    case 'green':
      return '#EFA300';
    case 'purple':
      return '#650000';
    case 'red':
      return '#B11515';
    default:
      return color;
  }
};
