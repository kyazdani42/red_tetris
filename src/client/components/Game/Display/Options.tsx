import * as React from 'react';
import { connect } from 'react-redux';

import { setOptions } from '../../../actions/actions';
import { State } from '../../../store';
import { OptionLabelWrapper, OptionsWrapper } from './styles';

interface Props {
  options: Options;
  dispatchSetOptions: (options: Options) => void;
}

const handleDispatch = (params: Props) => (option: { name: string; on: boolean }) =>
  params.dispatchSetOptions({ ...params.options, [option.name]: option.on });

export const Options: React.SFC<Props> = ({ options, dispatchSetOptions }) => {
  const dispatch = handleDispatch({ options, dispatchSetOptions });
  return (
    <OptionsWrapper>
      {checkbox(options.invisible, 'invisible', dispatch)}
      {checkbox(options.mirror, 'mirror', dispatch)}
      {checkbox(options.reverse, 'reverse', dispatch)}
    </OptionsWrapper>
  );
};

const checkbox = (
  checked: boolean,
  name: string,
  dispatch: (option: { name: string; on: boolean }) => void
) => (
  <OptionLabelWrapper>
    <label>{name}</label>
    <input type="checkbox" checked={checked} onChange={() => dispatch({ name, on: !checked })} />
  </OptionLabelWrapper>
);

const mapStateToProps = (state: State) => ({
  options: state.app.options
});

const mapDispatchToProps = (dispatch: any) => ({
  dispatchSetOptions: (options: Options) => dispatch(setOptions(options))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Options);
