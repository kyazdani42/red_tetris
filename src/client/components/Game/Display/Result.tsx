import * as React from 'react';

import { ResultStyle } from './styles';

interface Props {
  winner: boolean;
}
export const Result: React.SFC<Props> = ({ winner }) => (
  <ResultStyle winner={winner}>{winner ? 'You Win!' : 'Game Over!'}</ResultStyle>
);
