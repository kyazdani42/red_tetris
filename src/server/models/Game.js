const { timeout, generate } = require('../utils/game');
const Player = require('./Player');
const { removeGame, getGames } = require('../services/games');

module.exports = class Game {
  constructor({ name, io }) {
    this.io = io;
    this.name = name;
    this.allPieces = generate();
    this.owner = undefined;
    this.players = [];
    this.running = false;
    this.socket = this.initSocket(name, io);
  }

  initSocket(name, io) {
    const game = this;
    return io.of(`/${name}`).on('connection', socket => {
      if (game.running) {
        socket.disconnect();
      }
      game.addPlayer(socket);
      socket.on('disconnect', () => {
        game.removePlayer(socket.id);
      });
      socket.on('rotate', () => {
        game.actions(socket.id, 'rotate');
      });
      socket.on('moveDown', () => {
        game.actions(socket.id, 'moveDown');
      });
      socket.on('moveLeft', () => {
        game.actions(socket.id, 'moveLeft');
      });
      socket.on('moveRight', () => {
        game.actions(socket.id, 'moveRight');
      });
      socket.on('goDown', () => {
        game.actions(socket.id, 'goDown');
      });
      socket.on('start', () => {
        game.start(socket.id);
      });
      socket.on('stop', () => {
        console.log('stop');
        game.stop();
      });
    });
  }

  getPublicInfo() {
    return {
      name: this.name,
      running: this.running,
      // need the following 
      // players: this.players, // or .length,
      // owner: this.owner // owner name as we discussed, otherwise i might event don't need that
    };
  }

  privateInfo(player) {
    return {
      name: this.name,
      running: this.running,
      isOwner: player.id === this.owner,
      stack: player.tmpStack(),
      isPlaying: player.isPlaying,
    }
  }

  updateGame() {
    for (const player of this.players) {
      player.socket.emit('updateGame', this.privateInfo(player));
      console.log('updateGame')
    }
  }

  async run() {
    // await timeout();
    // this.socket.emit('updateData', {gameStatus: '3'});
    // await timeout();
    // this.socket.emit('updateData', {gameStatus: '2'});
    // await timeout();
    // this.socket.emit('updateData', {gameStatus: '1'});
    for (const player of this.players) {
      player.setNextPiece(this.allPieces[0]);
      player.isPlaying = true;
    }
    while (this.running) {
      await timeout();
      for (const player of this.players) {
        if (player.isPlaying) {
          const nbLine = player.updateStack();
          for (const looserPlayer of this.players) {
            if (looserPlayer.id !== player.id) {
              looserPlayer.addLine(nbLine);
            }
          }
          if (player.piece.fixed) {
            // allPieces peut etre vide, check et generer de nouvelle.
            player.setNextPiece(this.allPieces[player.pieceIndex]);
          }
          player.tryMoveDown();
        }
      }
      this.updateGame();
    }
  }

  start(id) {
    if (id === this.owner && !this.running) {
      // map player playing: true
      console.log('start');
      this.running = true;
      this.run();
    }
  }

  stop() {
    this.running = false;
  }

  addPlayer(socket) {
    if (!this.owner) {
      this.owner = socket.id;
    }
    this.players.push(new Player(socket));
    this.io.emit('games', getGames());
  }

  actions(id, action) {
    const player = this.players.find(d => d.id === id);
    if (!player.isPlaying) { return; }
    switch (action) {
      case 'moveDown':
        player.tryMoveDown();
        break;
      case 'rotate':
        player.tryRotate();
        break;
      case 'moveRight':
        player.tryMoveRight();
        break;
      case 'moveLeft':
        player.tryMoveLeft();
        break;
      case 'goDown':
        player.tryGoDown();
        break;
    }
    this.updateGame();
  };

  removePlayer(id) {
    this.players = this.players.filter(d => d.id !== id);
    if (this.owner === id && this.players.length) {
      this.owner = this.players[0].id;
    } else {
      removeGame(this.name);
    }
    this.io.emit('games', getGames());
  }
};
