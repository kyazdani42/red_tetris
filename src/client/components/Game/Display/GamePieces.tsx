import * as React from 'react';
import styled from 'styled-components';

import { Block } from './Block';
import { GamePiecesContainer } from './styles';

interface Props {
  stack: GameProps['stack'];
}

const getDivsFromStack = (stack: GameProps['stack']) => {
  const container = Array(200);
  for (let i = 0; i < 20; i++) {
    for (let j = 0; j < 10; j++) {
      const containerIndex = +`${i}${j}`;
      container[containerIndex] = (
        <Block color={stack[i][j].color} key={'block' + containerIndex} />
      );
    }
  }
  return container;
};

export const GamePieces: React.SFC<Props> = ({ stack }) => {
  if (!stack.length) return null;
  return <GamePiecesContainer>{getDivsFromStack(stack)}</GamePiecesContainer>;
};
