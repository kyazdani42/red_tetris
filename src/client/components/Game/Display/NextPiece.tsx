import * as React from 'react';

import { NextBlock } from './Block';
import { PieceContainer } from './styles';

interface Props {
  nextPiece: {
    color: string;
    patterns: number[][][];
  } | null;
}

export const NextPiece: React.SFC<Props> = ({ nextPiece }) => (
  <PieceContainer>
    {getBlocksFromPiece(nextPiece)}
  </PieceContainer>
);

const getBlocksFromPiece = (nextPiece: Props['nextPiece']) => {
  if (!nextPiece) return null;
  const piece = [];
  const len = nextPiece.patterns[0].length;
  for (let i = 0; i < 4; i++) {
    const line = i < len ? nextPiece.patterns[0][i] : null;
    for (let j = 0; j < 4; j++) {
      const blockValue = line ? j < line.length ? line[j] : 0 : 0;
      const key = `block${i}-${j}-next`;
      if (blockValue) {
        piece.push(<NextBlock color={nextPiece.color} key={key} />);
      } else {
        piece.push(<NextBlock color="black" key={key} />);
      }
    }
  }
  return piece;
};
