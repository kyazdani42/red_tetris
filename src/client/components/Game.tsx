import * as React from 'react';
import { connect } from 'react-redux';

import { State } from '../store';
import { GamePieces } from './GamePieces';
import { NextPiece } from './NextPiece';
import { Score } from './Score';

interface Props {
  gameData: GameProps | null;
}

export const Game: React.SFC<Props> = ({ gameData }) => {
  let gamePieces, score, nextPiece;
  if (gameData) {
    gamePieces = <GamePieces stack={gameData.stack} />;
    score = gameData.score;
    nextPiece = gameData.nextPiece;
  } else {
    gamePieces = null;
    score = 0;
    nextPiece = null;
  }
  return (
    <React.Fragment>
      {gamePieces}
      <Score score={score} />
      <NextPiece nextPiece={nextPiece} />
    </React.Fragment>
  );
};
const mapStateToProps = ({ app }: State) => ({
  gameData: app.gameData
});

export default connect(mapStateToProps)(Game);
