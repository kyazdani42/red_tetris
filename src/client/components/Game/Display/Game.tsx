import * as React from 'react';
import { connect } from 'react-redux';

import { State } from '../../../redux/store';
import { GamePieces } from './GamePieces';
import { NextPiece } from './NextPiece';
import Options from './Options';
import { Result } from './Result';
import { Score } from './Score';
import StartButton from './StartButton';
import { GameStartInfoWrapper } from './styles';

interface Props {
  gameData: GameProps | null;
  socket: SocketIOClient.Socket;
  playerScore: BestScore | null;
}

export const Game: React.SFC<Props> = ({ gameData, socket, playerScore }) => {
  let gamePieces, score, nextPiece, startButton, result, options;
  const bestScore = playerScore ? playerScore.bestScore : 0;
  const isPlaying = gameData && gameData.isPlaying || false;
  if (gameData) {
    score = gameData.score;
    nextPiece = gameData.nextPiece;

    const { stack, running, isOwner, winner } = gameData;
    gamePieces = stack.length ? <GamePieces stack={stack} /> : null;
    startButton = running || !isOwner ? null : <StartButton socket={socket} />;
    options = running || !isOwner ? null : <Options />;
    result = winner === undefined || isPlaying ? null : <Result winner={winner} />;
  } else {
    score = 0;
    nextPiece = null;
    gamePieces = null;
    startButton = null;
    result = null;
    options = null;
  }
  return (
    <React.Fragment>
      {gamePieces}
      <Score score={score} bestScore={bestScore} />
      <NextPiece nextPiece={nextPiece} />
      {!isPlaying &&
      <GameStartInfoWrapper>
        {startButton}
        {options}
        {result}
      </GameStartInfoWrapper>
      }
    </React.Fragment>
  );
};

export const mapStateToProps = ({ app }: State) => ({
  gameData: app.gameData,
  playerScore: app.playerScore,
  socket: app.socket
});

export default connect(mapStateToProps)(Game);
