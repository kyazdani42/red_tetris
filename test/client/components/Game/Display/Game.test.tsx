// tslint:disable
import * as React from 'react';
import { shallow } from 'enzyme';

import { mapStateToProps, Game } from '../../../../../src/client/components/Game/Display/Game';

const props: any = {
  gameData: {
    name: '',
    running: true,
    isOwner: true,
    winner: true,
    isPlaying: true,
    score: 0,
    stack: [[{
      color: '',
      value: 0,
      fix: true,
    }]],
    nextPiece: {
      color: '',
      patterns: [
        [
          [0]
        ],
      ],
    },
    otherPlayers: [{
      name: '',
      spectre: [0],
    }],
  },
  socket: null
};

describe('testing the component', () => {
  it('smoke tests the component', () => {
    const wrapper = shallow(<Game {...props} />)
    expect(wrapper.update()).toMatchSnapshot();
  }) 
})

describe('testing mapStateToProps', () => {
  const state: any = {
    app: {
      gameData: 'gameData',
      socket: 'socket'
    }
  }
  it('test the mapping', () => {
    const expected = { gameData: 'gameData', socket: 'socket' };
    expect(mapStateToProps(state)).toEqual(expected);
  })
})
