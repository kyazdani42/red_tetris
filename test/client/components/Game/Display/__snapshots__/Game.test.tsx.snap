// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component smoke tests the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    gameData={
      Object {
        "isOwner": true,
        "isPlaying": true,
        "name": "",
        "nextPiece": Object {
          "color": "",
          "patterns": Array [
            Array [
              Array [
                0,
              ],
            ],
          ],
        },
        "otherPlayers": Array [
          Object {
            "name": "",
            "spectre": Array [
              0,
            ],
          },
        ],
        "running": true,
        "score": 0,
        "stack": Array [
          Array [
            Object {
              "color": "",
              "fix": true,
              "value": 0,
            },
          ],
        ],
        "winner": true,
      }
    }
    socket={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Unknown
          stack={
            Array [
              Array [
                Object {
                  "color": "",
                  "fix": true,
                  "value": 0,
                },
              ],
            ]
          }
        />,
        <Unknown
          bestScore={0}
          score={0}
        />,
        <Unknown
          nextPiece={
            Object {
              "color": "",
              "patterns": Array [
                Array [
                  Array [
                    0,
                  ],
                ],
              ],
            }
          }
        />,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "stack": Array [
            Array [
              Object {
                "color": "",
                "fix": true,
                "value": 0,
              },
            ],
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "bestScore": 0,
          "score": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "nextPiece": Object {
            "color": "",
            "patterns": Array [
              Array [
                Array [
                  0,
                ],
              ],
            ],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Unknown
            stack={
              Array [
                Array [
                  Object {
                    "color": "",
                    "fix": true,
                    "value": 0,
                  },
                ],
              ]
            }
          />,
          <Unknown
            bestScore={0}
            score={0}
          />,
          <Unknown
            nextPiece={
              Object {
                "color": "",
                "patterns": Array [
                  Array [
                    Array [
                      0,
                    ],
                  ],
                ],
              }
            }
          />,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "stack": Array [
              Array [
                Object {
                  "color": "",
                  "fix": true,
                  "value": 0,
                },
              ],
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "bestScore": 0,
            "score": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "nextPiece": Object {
              "color": "",
              "patterns": Array [
                Array [
                  Array [
                    0,
                  ],
                ],
              ],
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
